File: app/src/main/java/org/imozerov/streetartview/ui/model/ArtObjectUi.java
package org.imozerov.streetartview.ui.model;

import org.imozerov.streetartview.storage.model.RealmArtObject;
import org.imozerov.streetartview.storage.model.RealmAuthor;
import org.imozerov.streetartview.storage.model.RealmString;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by imozerov on 05.02.16.
 */
public class ArtObjectUi {
    public final String id;
    public final String name;
    public final AuthorUi author;
    public final String description;
    public final String thumbPicUrl;
    public final List<String> picsUrls;

    public ArtObjectUi(String aId,
                       String aName,
                       AuthorUi aAuthor,
                       String aDescription,
                       String aThumbPicUrl,
                       List<String> aPicsUrls) {
        id = aId;
        name = aName;
        author = aAuthor;
        description = aDescription;
        thumbPicUrl = aThumbPicUrl;
        picsUrls = aPicsUrls;
    }

    public ArtObjectUi(RealmArtObject realmArtObject) {
        RealmAuthor realmAuthor = realmArtObject.getAuthor();

        List<String> picUrls = new ArrayList<>();
        for (RealmString realmString : realmArtObject.getPicsUrls()) {
            picUrls.add(realmString.getValue());
        }

        this.id = realmArtObject.getId();
        this.name = realmArtObject.getName();
        this.author =  new AuthorUi(realmAuthor);
        this.description = realmArtObject.getDescription();
        this.thumbPicUrl = realmArtObject.getThumbPicUrl();
        this.picsUrls = picUrls;
    }
}

File: app/src/main/java/org/imozerov/streetartview/ui/model/ArtObjectUi.kt
package org.imozerov.streetartview.ui.model

import android.support.annotation.VisibleForTesting
import org.imozerov.streetartview.storage.model.RealmArtObject
import java.util.*

/**
 * Created by imozerov on 05.02.16.
 */
class ArtObjectUi {
    val id: String
    val name: String
    val authors: List<AuthorUi>
    val description: String
    val thumbPicUrl: String
    val picsUrls: List<String>
    val lat: Double
    val lng: Double
    val address: String

    constructor(realmArtObject: RealmArtObject) {
        val picUrls = ArrayList<String>()
        for (realmString in realmArtObject.picsUrls) {
            picUrls.add(realmString.value)
        }

        id = realmArtObject.id
        name = realmArtObject.name
        authors = realmArtObject.authors.map {
            return@map AuthorUi(
                    it.id,
                    it.name,
                    it.photo
            )
        }
        description = realmArtObject.description
        thumbPicUrl = realmArtObject.thumbPicUrl ?: ""
        picsUrls = picUrls

        lat = realmArtObject.location.lat
        lng = realmArtObject.location.lng
        address = realmArtObject.location.address
    }

    @VisibleForTesting
    constructor(id: String, name: String, authors: List<AuthorUi>, description: String, thumbPicUrl: String, picsUrls: List<String>, lat: Double, lng: Double, address: String) {
        this.id = id
        this.name = name
        this.authors = authors
        this.description = description
        this.thumbPicUrl = thumbPicUrl
        this.picsUrls = picsUrls
        this.lat = lat
        this.lng = lng
        this.address = address
    }

    fun authorsNames() : String {
        return authors.map{ return@map it.name }.reduce { s1, s2 -> s1 + ", " + s2 }
    }

    fun matches(query: String) : Boolean {
        if (query.isEmpty()) {
            return true
        }

        if (name.toLowerCase().contains(query.toLowerCase())) {
            return true
        }

        if (authorsNames().toLowerCase().contains(query.toLowerCase())) {
            return true
        }

        return false
    }
}
